pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentMessageExpiryMonitor.java

```diff
@@ -78,10 +86,10 @@ public boolean expireMessages(int messageTTLInSeconds) {
         if (expirationCheckInProgressUpdater.compareAndSet(this, FALSE, TRUE)) {
             log.info("[{}][{}] Starting message expiry check, ttl= {} seconds", topicName, subName,
                     messageTTLInSeconds);
-
+            Map<Long, MLDataFormats.ManagedLedgerInfo.LedgerInfo> ledgerIdToInfoCache = new ConcurrentHashMap<>();
             cursor.asyncFindNewestMatching(ManagedCursor.FindPositionConstraint.SearchActiveEntries, entry -> {
                 try {
-                    long entryTimestamp = Commands.getEntryTimestamp(entry.getDataBuffer());
+                    long entryTimestamp = getEntryTimestamp(entry, ledgerIdToInfoCache, cursor);
```

Comment:
Is it better to split the message expiration into 2 steps?

1. Expire the closed ledgers if `ledger close time < current time - ttl`
2. Expire the entries (the existing method)

So that we don't need to touch the existing expire entries logic and this way will be more efficient since we don't need to read entries if the whole ledger is expired.